src/IO/WriteBufferFromS3.cpp

```diff
@@ -364,10 +364,16 @@ void WriteBufferFromS3::allocateBuffer()
 
 void WriteBufferFromS3::allocateFirstBuffer()
 {
+    /// First buffer was already allocated in BufferWithOwnMemory constructor with provided in constructor buffer size.
+    /// It will be reallocated in subsequent nextImpl calls up to the desired buffer size from buffer_allocation_policy.
+    /// But it may happen that buffer size provided in constructor is larger then desired buffer size from buffer_allocation_policy.
+    /// Resize memory in this case to the desired size.
     const auto max_first_buffer = buffer_allocation_policy->getBufferSize();
-    const auto size = std::min(size_t(DBMS_DEFAULT_BUFFER_SIZE), max_first_buffer);
-    memory = Memory(size);
-    WriteBuffer::set(memory.data(), memory.size());
+    if (memory.size() > max_first_buffer)
+    {
+        memory.resize(max_first_buffer);
+        WriteBuffer::set(memory.data(), memory.size());
+    }
```

Comment:
Previously we just allocated memory in constructor with `buf_size` provided in sonstructor and then allocated new memory here again.

Also, in `MergeTreeDataPartWriterOnDisk` we use `marks_compress_block_size` as block size for marks and it's 64KB by default, but in WriteBufferFromS3 we will eventually allocate much larger buffer because of buffer_allocation_policy. AFAIU we cannot avoid it for fixed-size mumultipart upload, but can we respect provided in constructor `buf_size` for other cases?

Reply:
Lets remove this function.
and rewrite this:
- `: WriteBufferFromFileBase(buf_size_, nullptr, 0)`
+ `: WriteBufferFromFileBase(std::min(buf_size_, DBMS_DEFAULT_BUFFER_SIZE), nullptr, 0)`

This would work as you described.